File: scala/Practice2/Others/Illumination.scala
// https://atcoder.jp/contests/joi2013ho/tasks/joi2013ho1

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    Array(List.empty[Tuple2[String, Int]]) |> (line => (
      io.StdIn.readLine |> (_ => io.StdIn.readLine.split(" ").foreach(x =>
        line(0) match {
          case (y, size)::ts if x != y => line(0) = (x, size + 1)::ts
          case _ => line(0) = (x, 1)::line(0)
        }
      ))
      |> (_ => line(0)
        .foldLeft((0, 0, 0)) { case ((x1, x2, acc), (_, x0)) =>
          (x0, x1, (x0 + x1 + x2).max(acc))
        }
      )
      |> { case (_, _, acc) => acc }
    ))
    |> println
  )
}

// https://atcoder.jp/contests/joi2013ho/submissions/28716573

File: fs/Practice2/Others/Illumination.fs
// https://atcoder.jp/contests/joi2013ho/tasks/joi2013ho1

[|[]|]
|> fun line ->
  stdin.ReadLine()
  |> fun _ -> stdin.ReadLine().Split()
  |> Array.iter (fun x ->
    match line.[0] with
    | (y, size)::ts when x <> y -> line.[0] <- (x, size + 1)::ts
    | _ -> line.[0] <- (x, 1)::line.[0]
  )
  |> fun () ->
    line.[0]
    |> List.fold (fun (x1, x2, acc) (_, x0) ->
      (x0, x1, max acc (x0 + x1 + x2))
    ) (0, 0, 0)
    |> fun (_, _, acc) -> acc
|> printfn "%A"

// https://atcoder.jp/contests/joi2013ho/submissions/28716389
